Snowflake.Data/Core/Authenticator/IAuthenticator.cs

```diff
@@ -0,0 +1,69 @@
+﻿/*
+ * Copyright (c) 2012-2019 Snowflake Computing Inc. All rights reserved.
+ */
+
+using System.Threading;
+using System.Threading.Tasks;
+using Snowflake.Data.Client;
+using Snowflake.Data.Log;
+
+namespace Snowflake.Data.Core.Authenticator
+{
+    /// <summary>
+    /// Interface for Authenticator
+    /// For simplicity, only the Asynchronous function call is created
+    /// </summary>
+    internal interface IAuthenticator
+    {
+        /// <summary>
+        /// Process the authentication asynchronouly
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        /// <exception cref="SnowflakeDbException"></exception>
+        Task AuthenticateAsync(CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Types of authenticators
+    /// </summary>
+    internal enum SFAuthenticatorType
+    {
+        SNOWFLAKE,
+        OKTA,
+    }
+
+    /// <summary>
+    /// Authenticator Factory to build authenticators
+    /// </summary>
+    internal class AuthenticatorFactory
+    {
+        private static readonly SFLogger logger = SFLoggerFactory.GetLogger<AuthenticatorFactory>();
+        /// <summary>
+        /// Generate the authenticator given the session
+        /// </summary>
+        /// <param name="session">session that requires the authentication</param>
+        /// <returns>authenticator</returns>
+        /// <exception cref="SnowflakeDbException">when authenticator is unknown</exception>
+        internal static IAuthenticator GetAuthenticator(SFSession session)
+        {
+            string type = session.properties[SFSessionProperty.AUTHENTICATOR];
+            switch (type)
```

Comment:
Maybe use if/elseif, switch seems not good here since you cannot switch for all cases.